### PR TITLE
Generator active power by region

### PR DESCRIPTION
**Pull request by AAJELLAL**

### CODE REVIEW

File: src/main/java/org/gridsuite/modification/server/modifications/GenerationDispatch.java [hunk 1]
```diff
@@ -522,13 +523,65 @@ public void apply(Network network, Reporter subReporter) {
...
         }
     }
+
+    private Map<String, List<Generator>> getGeneratorsByRegion(Network network) {
+        // get all connected generators and the substationIds associated.
+        List<Generator> connectedGenerators = network.getGeneratorStream()
+                .filter(g -> g.getTerminal().isConnected())
+                .toList();
+        List<String> substationIds = connectedGenerators.stream()
+                .map(g -> g.getTerminal().getVoltageLevel().getSubstation().orElseThrow().getId())
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/modification/server/modifications/GenerationDispatch.java:546]: Don't throw an exception for a voltageLevel not in a substation, which can occur : just ignore the generator in that case

File: src/main/java/org/gridsuite/modification/server/modifications/GenerationDispatch.java [hunk 2]
```diff
@@ -522,13 +523,65 @@ public void apply(Network network, Reporter subReporter) {
...
+
+        // group substationIds by region
+        Map<String, List<String>> groupedSubstationIds = substationIdPropertiesMap.keySet().stream().collect(Collectors.groupingBy(substationIdPropertiesMap::get));
+
+        // iterate over groupedSubstationIds and check for each substation list if it's related to the connected generators
+        Map<String, List<Generator>> generatorsByRegion = new HashMap<>();
+
+        groupedSubstationIds.forEach((region, substationList) -> {
+            List<Generator> connectedGeneratorsWithSubstation = connectedGenerators.stream()
+                    .filter(g -> substationList.contains(g.getTerminal().getVoltageLevel().getSubstation().orElseThrow().getId()))
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/modification/server/modifications/GenerationDispatch.java:571]: Same : Don't throw an exception for a voltageLevel not in a substation, which can occur : just ignore the generator in that case

File: src/main/java/org/gridsuite/modification/server/modifications/GenerationDispatch.java [hunk 3]
```diff
@@ -522,13 +523,65 @@ public void apply(Network network, Reporter subReporter) {
...
+                        Map.of("region", region, "sum", getSumActivePower(generators)), TypedValue.INFO_SEVERITY));
             } else {
                 double remainingPowerImbalance = totalAmountSupplyToBeDispatched - realized;
                 report(resultReporter, Integer.toString(componentNum), "SupplyDemandBalanceCouldNotBeMet", "The supply-demand balance could not be met : the remaining power imbalance is ${remainingPower} MW",
                     Map.of("remainingPower", remainingPowerImbalance), TypedValue.WARN_SEVERITY);
             }
         }
     }
+
+    private Map<String, List<Generator>> getGeneratorsByRegion(Network network) {
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/modification/server/modifications/GenerationDispatch.java:540]: You should add also the synchronous component to this method arguments, and filter the generators that are inside this synchronous component (see computeTotalAmountFixedSupply method for example)

**Review by FranckLecuyer — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/modification/server/modifications/GenerationDispatch.java (hunk 2 above)
Comment (AAJELLAL) [reply src/main/java/org/gridsuite/modification/server/modifications/GenerationDispatch.java]: Done.

File: src/main/java/org/gridsuite/modification/server/modifications/GenerationDispatch.java (hunk 3 above)
Comment (AAJELLAL) [reply src/main/java/org/gridsuite/modification/server/modifications/GenerationDispatch.java]: Done.

File: src/main/java/org/gridsuite/modification/server/modifications/GenerationDispatch.java (hunk 1 above)
Comment (AAJELLAL) [reply src/main/java/org/gridsuite/modification/server/modifications/GenerationDispatch.java]: Done.

**Review by FranckLecuyer — APPROVED** [review]

**Review by ayolab — APPROVED** [review]

**Pull request merged**